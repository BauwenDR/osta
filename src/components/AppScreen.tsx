import { IonBackButton, IonButtons, IonContent, IonHeader, IonPage, IonTitle, IonToolbar } from '@ionic/react'
import React, { ReactNode } from 'react'

interface PageProperties {
  name: string
  contentPadding?: boolean
  children?: ReactNode
}

export function AppScreen({ name, contentPadding = true, children }: PageProperties) {
  return (
    <IonPage>
      <IonHeader>
        <IonToolbar>
          <IonButtons slot="start">
            <IonBackButton />
          </IonButtons>
          <IonTitle>{name}</IonTitle>
        </IonToolbar>
      </IonHeader>
      <IonContent className={contentPadding ? 'ion-padding' : ''}>{children}</IonContent>
    </IonPage>
  )
}
