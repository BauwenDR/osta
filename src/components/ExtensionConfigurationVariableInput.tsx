import React from 'react'
import { IonInput, IonSelect, IonSelectOption } from '@ionic/react'
import { ExtensionConfigurationVariable } from '../types'
import { TextFieldTypes } from '@ionic/core'

interface ExtensionConfigurationVariableInputProperties {
  configurationVariable: ExtensionConfigurationVariable
  value?: string
  onInputChange: (value: string) => void
}

export function ExtensionConfigurationVariableInput({
  configurationVariable,
  value,
  onInputChange,
}: ExtensionConfigurationVariableInputProperties) {
  const { title, type, options, required } = configurationVariable

  if (type === 'select' && options && options.length > 0) {
    return (
      <IonSelect label={'Pick a ' + title} value={value} onIonChange={(event) => onInputChange(event.target.value)}>
        {options.map((option, index) => (
          <IonSelectOption key={index} value={option.value}>
            {option.name}
          </IonSelectOption>
        ))}
      </IonSelect>
    )
  } else {
    return (
      <IonInput
        required={required}
        type={type as TextFieldTypes}
        label={title}
        value={value}
        labelPlacement={'floating'}
        onIonChange={(event) => onInputChange(String(event.target.value))}
      />
    )
  }
}
