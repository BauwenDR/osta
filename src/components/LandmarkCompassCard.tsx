import React from 'react'
import { Landmark } from '../types'
import { IonCard, IonCardContent, IonCardHeader, IonCardTitle, IonImg } from '@ionic/react'
import LocationHelper from '../helpers/LocationHelper'
import { useCompassDirection } from '../hooks/useCompassDirection'
import { useLocation } from '../hooks/useLocation'

import './css/LandmarkCompassCard.css'

interface LandmarkCompassCardProperties {
  landmark: Landmark
  currentLandmarkNumber: number
  totalLandmarks: number
}

export function LandmarkCompassCard({ landmark, currentLandmarkNumber, totalLandmarks }: LandmarkCompassCardProperties) {
  const location = useLocation()
  const userDirection = useCompassDirection()

  const distanceToNextPoint = LocationHelper.getDisplayStringForDistanceInKm(landmark.location, location)
  const rotationToNextPoint = (-(LocationHelper.calculateAngleInDegrees(landmark.location, location) + userDirection) % 360) + 180

  const title = landmark.sources[0].name

  return (
    <IonCard className="compassCard">
      <div className="compassCard__text">
        <IonCardHeader>
          <IonCardTitle>
            {title.length > 20 ? (
              <>
                Next stop: {title.substring(0, 20)}... ({currentLandmarkNumber}/{totalLandmarks})
              </>
            ) : (
              <>
                Next stop: {title}... ({currentLandmarkNumber}/{totalLandmarks})
              </>
            )}
          </IonCardTitle>
        </IonCardHeader>
        <IonCardContent>Remaining distance: {distanceToNextPoint}</IonCardContent>
      </div>
      <div className="compassCard__image image__light">
        <IonImg
          style={{
            transformOrigin: '50% 50%',
            transform: `rotate(${rotationToNextPoint}deg)`,
          }}
          src="/images/compass/compass.svg"
        />
      </div>
      <div className="compassCard__image ion-palette-dark image__dark">
        <IonImg
          style={{
            transformOrigin: '50% 50%',
            transform: `rotate(${rotationToNextPoint}deg)`,
          }}
          src="/images/compass/compass-dark.svg"
        />
      </div>
    </IonCard>
  )
}
