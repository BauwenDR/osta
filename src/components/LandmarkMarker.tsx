import React from 'react'
import { Marker, Popup } from 'react-leaflet'
import { IonRow, IonText, IonButton, IonIcon } from '@ionic/react'
import { informationCircleOutline } from 'ionicons/icons'
import { Landmark } from '../types'
import L from 'leaflet'

interface LandmarkMarkerProps {
  landmark: Landmark
  markerIcon: L.Icon
  markerIconVisited: L.Icon
}

export function LandmarkMarker({ landmark, markerIcon, markerIconVisited }: LandmarkMarkerProps) {
  return (
    <Marker position={[landmark.location.lat, landmark.location.lng]} icon={landmark.visited ? markerIconVisited : markerIcon}>
      <Popup>
        <IonRow>
          <IonText className={'popup__text__center'}>
            <IonButton className={'btn__color'} routerLink={`/landmark/${landmark.id}`}>
              <IonIcon icon={informationCircleOutline} />
            </IonButton>
            <IonText>{landmark.sources[0].name}</IonText>
          </IonText>
        </IonRow>
      </Popup>
    </Marker>
  )
}
