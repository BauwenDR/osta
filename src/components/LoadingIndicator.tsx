import React from 'react'
import { IonGrid, IonImg, IonRow, IonSpinner, IonText } from '@ionic/react'

import './css/StatusScreen.css'

interface LoadingScreenProps {
  text: string
}

export function LoadingIndicator({ text }: LoadingScreenProps) {
  return (
    <IonGrid className="statusContainer">
      <IonRow>
        <IonImg className="statusContainer__image" src="/images/loading.png"></IonImg>
      </IonRow>
      <IonRow>
        <IonSpinner color="primary" />
      </IonRow>
      <IonRow>
        <IonText className="statusContainer__text">{text}</IonText>
      </IonRow>
      <IonRow>
        <IonText className="statusContainer__credentials">This image was generated by Microsoft CoPilot.</IonText>
      </IonRow>
    </IonGrid>
  )
}
