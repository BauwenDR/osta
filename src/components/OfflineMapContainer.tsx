import L, { Map } from 'leaflet'
import React, { useEffect, useRef } from 'react'
import { MapContainer } from 'react-leaflet'
import 'leaflet.offline'
import { Location } from '../types'
import { IonFabButton, IonIcon } from '@ionic/react'
import { navigateOutline } from 'ionicons/icons'

import 'leaflet/dist/leaflet.css'
import './css/LeafletMap.css'
import 'leaflet-draw/dist/leaflet.draw.css'

interface OfflineMapContainerProperties {
  center: Location
  zoom: number
  scrollWheelZoom: boolean
  children?: React.ReactNode
  className?: string
}

export function OfflineMapContainer({ center, zoom, scrollWheelZoom, children, className }: OfflineMapContainerProperties) {
  const map = useRef<Map | null>(null)

  useEffect(() => {
    if (map.current) {
      // @ts-expect-error: leaflet.offline is a js only library, thus js cannot find the function even though it exists
      const tileOfflineLayer = L.tileLayer.offline('https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png', {
        attribution: '&copy; <a href="https://osm.org/copyright">OpenStreetMap</a> contributors',
      })
      tileOfflineLayer.addTo(map.current)

      map.current.on('load', () => {
        map.current?.invalidateSize()
      })
    }
  }, [map])

  function handleCenterClick() {
    if (map.current) {
      map.current.flyTo(center, zoom)
    }
  }

  return (
    <>
      <MapContainer
        ref={map}
        center={[center.lat, center.lng]}
        zoom={zoom}
        maxZoom={18}
        minZoom={4}
        scrollWheelZoom={scrollWheelZoom}
        className={className}
        worldCopyJump={true}
      >
        {children}
      </MapContainer>
      <IonFabButton color="light" onClick={handleCenterClick} className="btn__home btn__center">
        <IonIcon icon={navigateOutline}></IonIcon>
      </IonFabButton>
    </>
  )
}
