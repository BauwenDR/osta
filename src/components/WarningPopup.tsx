import React from 'react';
import { IonAlert } from '@ionic/react';

interface ErrorWarningProps {
  title: string
  message: string
  isOpen: boolean
}

export function WarningPopup({ title, message, isOpen}: ErrorWarningProps) {
  return (
      <IonAlert
        isOpen={isOpen}
        header={title}
        message={message}
        buttons={['OK']}
      />
  )
}
