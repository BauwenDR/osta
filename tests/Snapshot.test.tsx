import React from 'react'
import renderer from 'react-test-renderer'
import { WarningPopup } from '../src/components/WarningPopup'
import { LoadingIndicator } from '../src/components/LoadingIndicator'

it('Error warning should render correctly', () => {

  const tree = renderer.create(<WarningPopup title="Warning" message="Something went wrong." isOpen={true} />).toJSON()
  expect(tree).toMatchSnapshot()
})

it('Loading indicator should render correctly', () => {
  const tree = renderer.create(<LoadingIndicator text="Loading information" />).toJSON()
  expect(tree).toMatchSnapshot()
})
