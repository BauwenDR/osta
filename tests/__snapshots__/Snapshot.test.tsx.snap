// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Error warning should render correctly 1`] = `
<template>
  <ion-alert
    header="Warning"
    is-open={true}
    message="Something went wrong."
  />
</template>
`;

exports[`Loading indicator should render correctly 1`] = `
<ion-grid>
  <ion-row>
    <ion-img
      src="/images/loading.png"
    />
  </ion-row>
  <ion-row>
    <ion-spinner
      color="primary"
    />
  </ion-row>
  <ion-row>
    <ion-text>
      Loading information
    </ion-text>
  </ion-row>
  <ion-row>
    <ion-text>
      This image was generated by Microsoft CoPilot.
    </ion-text>
  </ion-row>
</ion-grid>
`;
